flask/flask-launcher/src/main/java/net/corda/flask/launcher/Launcher.java

```diff
@@ -152,18 +154,40 @@ final int launch(Manifest manifest, List<String> jvmArgs, List<String> javaAgent
             }
             builder.getCliArgs().addAll(args);
             beforeChildJvmStart(builder);
-            try(LockFile processLock = LockFile.acquire(cache.getPidFile(), false)) {
-                builder.getProperties().put(Flask.JvmProperties.PID_FILE, cache.getPidFile());
-                Path heartbeatAgentPath = extractedLibraries.get(
-                        manifest.getMainAttributes().getValue(Flask.ManifestAttributes.HEARTBEAT_AGENT_HASH));
-                builder.getJvmArgs().add("-javaagent:" + heartbeatAgentPath);
-                int returnCode = builder.exec();
-                afterChildJvmExit(returnCode);
-                cache.touchLibraries();
-                return returnCode;
-            } finally {
-                Files.delete(cache.getPidFile());
-            }
+            builder.getProperties().put(Flask.JvmProperties.PID_FILE, cache.getPidFile());
+            Path heartbeatAgentPath = extractedLibraries.get(
+                    manifest.getMainAttributes().getValue(Flask.ManifestAttributes.HEARTBEAT_AGENT_HASH));
+            builder.getJvmArgs().add("-javaagent:" + heartbeatAgentPath);
+            LockFile processLock = LockFile.acquire(cache.getPidFile(), false);
+            Process process = builder.build().inheritIO().start();
+            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                if(process.isAlive()) {
+                    process.destroy();
+                    try {
+                        long timeout = Long.parseLong(
+                                System.getProperty(
+                                        Flask.JvmProperties.KILL_TIMEOUT_MILLIS,
+                                        Flask.Constants.DEFAULT_KILL_TIMEOUT_MILLIS));
+                        process.waitFor(timeout, TimeUnit.MILLISECONDS);
+                    } catch (InterruptedException ie) {
+                        log.error(ie.getMessage(), ie);
+                    } finally {
+                        if (process.isAlive()) {
+                            process.destroyForcibly();
+                        }
+                        processLock.close();
+                    }
+                }
+                try {
+                    Files.delete(cache.getPidFile());
+                } catch (IOException ex) {
```

Comment:
Does this also need to be inside the `try-finally` block? An `InterruptedException` will currently skip these lines. Alternatively, you could provide the thread with an `UncaughtExceptionHandler`.,

Reply:
I do not think that `Files.delete` can ever throw `InterruptedException`, the goal was to not block the execution of the code if deletion of the pidfile fails (which doesn't make much sense atm since that statement is already the last statement in its thread). I will add `cache.touchLibraries();` after the deletion of the pidfile and change the exception type in the catch clause to `Exception`

Reply:
My bad - I was thinking that any `InterruptedException` thrown by `process.waitFor()` would propagate out of the handler without reaching `Files.delete` :facepalm:.

Reply:
this why I hate checked exceptions... they only reason I wrote that 
```
} catch (InterruptedException ie) {
```
line is because the compiler forced me to do so, I do not expect that exception to ever happen